### PR TITLE
BE-1255: Microsoft DevOps integration WIP - do not merge

### PR DESCRIPTION
**Pull request by mdiazitexico**
**Description:**  Implementation of Microsoft Azure DevOps. WIP
**Notes:**  This integration implements normal fetch functions for DevOps core information but is also registered within git integrations in an attempt of use the same standard as we use with other git integrations.

### CODE REVIEW

File: integrations/azure/api/http.go [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+
+	transport := httpdefaults.DefaultTransport()
+
+	return &AzureAPI{
+		logger: logger,
+		client: &http.Client{
+			Transport: transport,
+			Timeout:   10 * time.Minute,
+			// CheckRedirect: createRedirectPolicy(ctx),
+		},
```
Comment (jhaynie) [inline integrations/azure/api/http.go]: lets not leave in commented code

File: integrations/azure/api/http.go [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+}
+
+// FetchProjects will fetch all the azure projects for current user
+func (a *AzureAPI) FetchWorkItems(ctx context.Context, projectId string) ([]WorkItem, error) {
+	var idsParam string
+	wiIDs, err := a.FetchWorkItemsIDs(ctx, projectId)
+	if err != nil {
+		return nil, err
+	}
+	if len(wiIDs) <= 200 {
```
Comment (jhaynie) [inline integrations/azure/api/http.go]: what happens if >200 here?

File: integrations/azure/api/http.go [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return boardRes.Value, nil
+}
+
+type responseDecoderFunc func(resp *http.Response, reader io.Reader) (interface{}, error)
+
+func addAuthorizationForRequest(ctx context.Context, req *http.Request) error {
+
```
Comment (jhaynie) [inline integrations/azure/api/http.go]: do they only support base64 or is there oauth2 support as well?

File: integrations/azure/api/http.go [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+		return v.String(), nil
+	}
+	return "", fmt.Errorf("missing Azure organization in context")
+}
+
+func (a *AzureAPI) FetchSingleRepo(ctx context.Context, project string) (*git.Repository, error) {
+	var repoRes Repo
+	var repo git.Repository
+
+	repoid := "testrepoid" //this will be removed with getRepos implementation
```
Comment (jhaynie) [inline integrations/azure/api/http.go]: these look hard coded.

File: integrations/azure/api/http.go [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+func GetAzureClient() (*AzureAPI, context.Context, error) {
+	return azureHTTPClient, azureContext, nil
+}
+
+type refreshAuth struct {
+	logger      log.Logger
+	integration integration.Integration
+}
+
+func (a *refreshAuth) SetAuth(req *http.Request) error {
```
Comment (jhaynie) [inline integrations/azure/api/http.go]: seem redundant the above auth method

File: integrations/azure/azure.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+// Name is the integration key found in the configuration
+func (i *Integration) Name() string {
+	return integrationName
+}
+
+// IsGit is the integration git or not
+func (i *Integration) IsGit() bool {
+	return false
```
Comment (jhaynie) [inline integrations/azure/azure.go]: ?? isn't this suppose to be true?

File: integrations/azure/azure.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	logger = log.With(logger, "pkg", integrationName)
+	newctx, a, err := i.newAPI(ctx, logger)
+	if err != nil {
+		helper.Fail(i, "%v", err)
+		return
+	}
+	if !testLogin(newctx, logger, i, helper, a) {
+		return
+	}
+	// testProjects(newctx, logger, i, helper, a)
```
Comment (jhaynie) [inline integrations/azure/azure.go]: we should at least test 1 project if possible

File: integrations/azure/export.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+			projectDir := filepath.Join(rootdir, id)
+			if err := os.MkdirAll(projectDir, 0755); err != nil {
+				return fmt.Errorf("error creating directory %v. %v", projectDir, err)
+			}
+			// FETCHING WORK ITEMS PER PROJECT
+			wi, err := azureApi.FetchWorkItems(ctx, id)
+			if err != nil {
+				return fmt.Errorf("error fetching azure work items. %v", err)
+			}
+			fn = filepath.Join(rootdir, id+"/", "workItems.json.gz")
```
Comment (jhaynie) [inline integrations/azure/export.go]: lower case and make it `work_item`

File: integrations/azure/export.go [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+	FetchRepoCommits: func(repoName string, branchName string) ([]*git.Commit, error) {
+		return nil, nil //TODO
+	},
+	FetchPullRequests: func(repoName string) ([]*git.PullRequest, error) {
+		return nil, nil //TODO
+	},
+	GitCloneURL: func(reponame string, project string) (string, error) {
+		return "", nil //TODO
+	},
+	Exclude: regexp.MustCompile("^noreply@github\\.com$"), //TODO
```
Comment (jhaynie) [inline integrations/azure/export.go]: remove this github

File: integrations/azure/validate.go [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+)
+
+func testLogin(ctx context.Context, logger log.Logger, integration *Integration, helper integration.ValidateHelper, azure api.AzureAPI) bool {
+	// started := time.Now()
+	// user, err := azure.FetchCurrentUser(ctx)
+	// if err != nil {
+	// 	helper.Fail(integration, "error fetching Azure service account user. %v", err)
+	// 	return false
+	// }
+	// helper.Pass(integration, "Azure service account user %v (%v) verified (%v)", user.DisplayName, user.EmailAddress?, time.Since(started)?)
```
Comment (jhaynie) [inline integrations/azure/validate.go]: we need some sort of validation that we can connect and things work ok

File: integrations/azure/azure.go (hunk 6 above)
Comment (mdiazitexico) [reply integrations/azure/azure.go]: not exactly, I'll double check but I think this function is only used while we run the generic export process
https://github.com/pinpt/agent/blob/BE-1255/cmd/root.go#L310
for git export just consider the registered gitIntegrations
so to run both for now we implemented both ways and leaved this in false; we are planning maybe to change the behavior of that flag so we can have other cases but didn't want to change the general flow for other integrations for now

File: integrations/azure/api/http.go (hunk 3 above)
Comment (mdiazitexico) [reply integrations/azure/api/http.go]: They do support oauth2; we are using this one for now as we test/fix fetches but we are starting to work on that

File: integrations/azure/validate.go (hunk 10 above)
Comment (mdiazitexico) [reply integrations/azure/validate.go]: we are starting to work on that implementation

File: integrations/azure/api/http.go (hunk 3 above)
Comment (jhaynie) [reply integrations/azure/api/http.go]: ok makes sense

File: integrations/azure/azure.go (hunk 6 above)
Comment (jhaynie) [reply integrations/azure/azure.go]: ok

**Comment by jhaynie** [comment]
I added Azure OAuth2 support for our auth api so you can now go here to generate a token/refresh token:

https://auth-api.pinpt.io/oauth/azure

The token can be used like this:

`Authorization: Bearer <TOKEN>`

You can follow the other integrations (search for refresh) to refresh the token on a 401.

https://github.com/pinpt/agent/blob/europa/integrations/github/api/http.go#L147

You can use this configuration example:

https://github.com/pinpt/agent/blob/europa/integrations/bitbucket/bitbucket.go#L49

Notice we support either/or apitoken/refresh token or username/password. Same with how we set auth headers.

**Comment by mdiazitexico** [comment]
> I added Azure OAuth2 support for our auth api so you can now go here to generate a token/refresh token:

We'll work on that. Thanks a lot

**Comment by rommel-sanchez** [comment]
closing PR, clean commits in #180

**Pull request closed without merging**